planner/core/stats.go

```diff
@@ -299,6 +303,106 @@ func (p *LogicalJoin) deriveStats() (*property.StatsInfo, error) {
 	return p.stats, nil
 }
 
+func (p *LogicalJoin) deriveInnerJoinStatsWithHist(leftKeys, rightKeys []*expression.Column) (*property.StatsInfo, error) {
+	leftChild, rightChild := p.children[0], p.children[1]
+	leftProfile, rightProfile := leftChild.statsInfo(), rightChild.statsInfo()
+
+	cardinality := make([]float64, 0, p.schema.Len())
+	cardinality = append(cardinality, leftProfile.Cardinality...)
+	cardinality = append(cardinality, rightProfile.Cardinality...)
+
+	ndv, leftNdv, rightNdv := float64(1), float64(1), float64(1)
+	idxHistID := int64(0)
+	newColID2Hist := make(map[int64]*statistics.Column)
+	newIdxID2Hist := make(map[int64]*statistics.Index)
+	leftIndexReLabel := make(map[int64]int64)
+	rightIndexReLabel := make(map[int64]int64)
+	newIdx2ColumnIDs := make(map[int64][]int64)
+
+	// TODO: Support using index histogram to calculate the NDV after join and the final row count.
+	for i := range leftKeys {
+		leftHist, ok1 := leftChild.statsInfo().HistColl.Columns[leftKeys[i].UniqueID]
+		rightHist, ok2 := rightChild.statsInfo().HistColl.Columns[rightKeys[i].UniqueID]
+		lPos := leftChild.Schema().ColumnIndex(leftKeys[i])
+		rPos := rightChild.Schema().ColumnIndex(rightKeys[i])
+		leftNdv *= leftProfile.Cardinality[lPos]
+		rightNdv *= rightProfile.Cardinality[rPos]
+		if ok1 && ok2 {
+			newHist := statistics.MergeHistogramForInnerJoin(&leftHist.Histogram, &rightHist.Histogram, leftKeys[i].RetType)
+			leftCol := &statistics.Column{Info: leftHist.Info, Histogram: *newHist}
+			rightCol := &statistics.Column{Info: rightHist.Info, Histogram: *newHist}
+			ndv *= float64(newHist.NDV)
+			lPosNew := p.schema.ColumnIndex(leftKeys[i])
+			rPosNew := p.schema.ColumnIndex(rightKeys[i])
+			cardinality[lPosNew] = ndv
+			cardinality[rPosNew] = ndv
+			newColID2Hist[leftKeys[i].UniqueID] = leftCol
+			newColID2Hist[rightKeys[i].UniqueID] = rightCol
+			continue
+		}
+		keyNdv := math.Min(leftChild.statsInfo().Cardinality[lPos], rightChild.statsInfo().Cardinality[rPos])
+		ndv *= keyNdv
+	}
+	count := leftProfile.RowCount / leftNdv * rightProfile.RowCount / rightNdv * ndv
+
+	// Update left column map in `HistColl`.
+	for uniqID, colHist := range leftProfile.HistColl.Columns {
+		_, ok := newColID2Hist[uniqID]
+		if ok {
+			continue
+		}
+		newColID2Hist[uniqID] = colHist
+	}
+
+	// Update right column map in `HistColl`.
+	for uniqID, colHist := range rightProfile.HistColl.Columns {
+		_, ok := newColID2Hist[uniqID]
+		if ok {
+			continue
+		}
+		newColID2Hist[uniqID] = colHist
+	}
```

Comment:
When calculating row count by histogram such as `GetRowCountByColumnRanges`, tt will call `getIncreaseFactor` to adjust the result with HistColl's RowCount.
I think this can cover the other histograms?

Reply:
Hmm, for inner join, the rows of the join result may change compared with the left/right input, and we have no `getIncreaseFactor` function call in procedure?

Reply:
For example, if the query is `select t1.b, t2.b from t1 join t2 on t1.a = t2.a`, for `DataSource(t1)` and `DataSource(t2)`, the `Buckets.Count` of column `b`'s histogram may be like `1,3,7,10`, that is to say, `t1` and `t2` has 10 rows respectively, if the join result contains 3 rows in total, then would it be problematic if we copy `b`'s histogram from `DataSource` to `LogicalJoin`? because the `Buckets.Count` of column `b` is still `1,3,7,10`, while the total count of join result is 3.